rust_build_utils/msvc.py

```diff
@@ -0,0 +1,149 @@
+                mvs_path
+            )
+        )
+
+    # Multiple versions and multiple editions might be installed, so we iterate over versions installed.
+    # If version preference is given, we filter only matching versions, othervise the highest version is picked.
+    mvs_versions = sorted(
+        [
+            v
+            for v in mvs_path.iterdir()
+            if v.is_dir()
+            and (version_preference is None or v.name == version_preference)
+        ],
+        key=lambda p: p.name,
+        reverse=True,
+    )
+    if len(mvs_versions) == 0:
+        raise Exception(
+            "Microsoft Visual Studio version not found. Was looking in '{}'".format(
+                mvs_path
+            )
+        )
+    msv_version = mvs_versions[0]
+
+    # There can be multiple editions of visual studio installed, but we're choosing based on a preference list.
+    # To pick the edition based on preference list, we create a function that returns an index in the list.
+    # When used as a sort key, the first element will be the closes to the start of the list
+    # To support values not in a list, the length of the list is used as fallback,
+    # putting those values effectively at the end.
+    edition_preferences = [
+        "BuildTools",
+        "Enterprise",
+        "Professional",
+        "Community",
+        "Preview",
+    ]
+    preference_index = (
+        lambda e: edition_preferences.index(e)
+        if e in edition_preferences
+        else len(edition_preferences)
+    )
+    # if edition_preference is given, the list will only contain that edition.
+    msv_editions = sorted(
+        [
+            e
+            for e in msv_version.iterdir()
+            if e.is_dir()
+            and (edition_preference is None or e.name == edition_preference)
+        ],
+        key=lambda p: preference_index(p.name),
+    )
+    if len(msv_editions) == 0:
+        raise Exception(
+            "Microsoft Visual Studio edition not found. Was looking in '{}'".format(
+                msv_version
+            )
+        )
+    msv = msv_editions[0]
+    vcvarsall = msv.joinpath(r"VC\Auxiliary\Build\vcvarsall.bat")
+
+    # architecture string that will be passed to vcvarsall
+    if not direct_pass_arch and arch == "aarch64":
+        arch = "arm64"
+    arch = (
+        arch
+        if direct_pass_arch or arch in ("amd64", "x64")
+        else "amd64_{}".format(arch)
+    )
+
+    original_env = {}
+    try:
+        # Execute vcvarsall in a shell and print the environment after modification.
+        # Because the change happens in a separate process, after it exits the changes made to the env are lost.
+        # We collect the process output (modified environment) and set current environment to those values.
+        # When setting the environment we save the old values so they can be restored after exiting the context.
+        p = subprocess.run(

```

Comment:
This looks like it should be `subprocess.check_output`.

Reply:
Didn't do this change. The piped output would add noise and it's not needed. The output is parsed after the process exits.

Reply:
Can you try this? `subprocess.check_call(['your_command'], stderr=subprocess.DEVNULL)`

Reply:
https://docs.python.org/3/library/subprocess.html#subprocess.check_call:

> Code needing to capture stdout or stderr should use [run()](https://docs.python.org/3/library/subprocess.html#subprocess.run) instead: